### ISSUE TITLE
Roadmap: Riju architecture improvements

### ISSUE THREAD

**Issue by raxod502** [radian-software/riju#151]
Riju has some serious scalability and maintainability problems, and I've been thinking for the last few months about how to address them. For transparency, I created the following roadmap outlining my plans:

https://nextcloud.intuitiveexplanations.com/s/mHoFwFGMNCpLjim
[note: moved from Google Drive to Nextcloud]

Feedback is of course welcome!

**Comment by hello-smile6** [comment]
> Riju has some serious scalability and maintainability problems, and I've been thinking for the last few months about how to address them. For transparency, I created the following roadmap outlining my plans:
> 
> https://docs.google.com/document/d/e/2PACX-1vRKkmilDiHXW8wpwFCplLrXRo-ANCLLTPpJ7vl4pI4diZAxr56vyuAo6NQ0ZTVYhf-X5HcQlVx3Ud9u/pub
> 
> Feedback is of course welcome!

Don't have access to the Google Doc, could you share it as a PDF?

**Comment by LuckyHookin** [comment]
@hello-smile6 
[Roadmap_ Riju architecture improvements.pdf](https://github.com/raxod502/riju/files/7948855/Roadmap_.Riju.architecture.improvements.pdf)

**Comment by raxod502** [comment]
I'm curious what issue you ran into, I seem to be able to access the link in your comment using unauthenticated curl, without even any JavaScript?

**Comment by hello-smile6** [comment]
> I'm curious what issue you ran into, I seem to be able to access the link in your comment using unauthenticated curl, without even any JavaScript?

Permissions issues with non-gmail account.

**Comment by raxod502** [comment]
You probably saw the original version of the link, which was broken. I edited the issue shortly after posting it. Try it again.

**Comment by hello-smile6** [comment]
> You probably saw the original version of the link, which was broken. I edited the issue shortly after posting it. Try it again.

Okay

**Comment by hello-smile6** [comment]
> You probably saw the original version of the link, which was broken. I edited the issue shortly after posting it. Try it again.

![image](https://user-images.githubusercontent.com/73048226/151724469-63a83352-f38d-4583-99ff-4c1375359bfb.png)
Yes, I removed my email.

**Comment by raxod502** [comment]
Huh, not sure what to say. When signed into another Google account that doesn't have access to that document, it works for me. Also works if I'm not signed into any Google account. It's not an IP address restriction either since I can even access it through Tor. And, as I mentioned, `curl` works fine with no authentication whatsoever. Perhaps it's some kind of issue that affects your account specifically. Seems like a pretty rare issue judging by the lack of discussion online about the error message.

**Comment by hello-smile6** [comment]
> Huh, not sure what to say. When signed into another Google account that doesn't have access to that document, it works for me. Also works if I'm not signed into any Google account. It's not an IP address restriction either since I can even access it through Tor. And, as I mentioned, `curl` works fine with no authentication whatsoever. Perhaps it's some kind of issue that affects your account specifically. Seems like a pretty rare issue judging by the lack of discussion online about the error message.

Okay, maybe use this link? It goes through a CORS proxy.
https://cors-anywhere.9pfs.repl.co/docs.google.com/document/d/e/2PACX-1vRKkmilDiHXW8wpwFCplLrXRo-ANCLLTPpJ7vl4pI4diZAxr56vyuAo6NQ0ZTVYhf-X5HcQlVx3Ud9u/pub

**Comment by hello-smile6** [comment]
@raxod502 I just used Tor as a test (for the first time) to open it, and I can confirm that it works.

**Comment by plondon** [comment]
Wow this is a super exciting proposal!

**Comment by raxod502** [comment]
Prometheus metrics are :fire: and much better than CloudWatch.

![image](https://user-images.githubusercontent.com/6559064/153699407-9ba2b875-41ac-437d-bb56-f102253f7edc.png)

**Comment by hello-smile6** [comment]
Did you have an outage at 21:55 in that screenshot?

**Comment by raxod502** [comment]
No, this is all on a staging instance not serving any live traffic.

**Comment by hello-smile6** [comment]
> No, this is all on a staging instance not serving any live traffic.

Oh.

**Comment by plondon** [comment]
Hi @raxod502!

Checking in on the progress and timeline (if any) of this roadmap?

**Comment by raxod502** [comment]
Hey @plondon. Sorry for the delay - I've moved apartments and have been spending the last few weeks unpacking boxes and putting together furniture.

I haven't done much work on this specific roadmap, mostly because I've had other ideas about how to rebuild things from scratch using the knowledge I've gained over the last year from other infrastructure-related personal projects as well as my experience operating Kubernetes at Plaid.

I have a project in progress that is a Heroku-like platform with true user authentication and workload scheduling on Kubernetes, and a separate project that rewrites Riju to use Go microservices and a relational database. Neither of those are MVP enough that I've put them up on GitHub yet. Learning a bunch of things, but since these are all things I'm working on ad hoc in my free time, they compete with all the other stuff going on.

Let me know if I can answer any specific questions!

**Comment by nyabinary** [comment]
What the progress on this so far now?

**Comment by raxod502** [comment]
Unfortunately, still https://github.com/radian-software/riju/issues/168. Considerable architectural work is needed to get Riju back into a state where I am able to deploy changes, and I haven't had time yet with all the other things that have taken time in my personal computing and non-computing life.
